### DISCUSSION TITLE
Is it time to consider React.js vs jQuery

### DISCUSSION THREAD

**Discussion by dbltoe** [zencart/zencart#5958] [General]
Chrome's Lighthouse seems to constantly point out aged/vulnerable jQuery installations and Lighthouse is blaming it for bloating,  Is it time to consider something like React.js?

https://www.makeuseof.com/jquery-react-migrate-app/#:~:text=There%20are%20several%20reasons%20why,representation%20of%20the%20actual%20DOM.

Their Why Migrate From jQuery to React? seems to make some sense.

Still, since it's not really my area of expertise, I thought it should at least be discussed.

**Comment by drbyte** [comment]
> If you're looking for a library that is easy to use and has a large community, then jQuery is a good choice. But if you're looking for a library that is more suited for large-scale development, React is the better option.

React is more for a situation where you can code for a specific use-case. Requires programming skills and understanding and tech-stack (software dev tools etc). Usually for building a specific app. Definitely much harder for unskilled end-users to customize/maintain/manage/upgrade.

jQuery tends to lend itself to drop-in functionality, enablement of plugins in specific situations as needed. Doesn't require compiling the app. Much more friendly to the casual plugin developer and easier to maintain for unskilled end-users (storeowners).

**Reply by dbltoe** [reply]
I figured there might be something in there that wasn't familiar enough we me to make a qualified assessment.  And, I should have known you were on top of it already. 😃

**Comment by zcwilt** [comment]
Why not Vue3 or angular or nuxt or alpine.js etc etc, why React

For all of these there is a development process, 
needing to run npm install and npm run build which adds an extra burden to the development process 
personally I'm not against that, as it is how I work in my day job.

However currently we are trying to make the dev process for Zen Cart as simple as possible.

**Comment by webconcept-one** [comment]
jQuery is used at zen-cart mainly for image size matching, right?
In my case the Front-Page is loaded fully after 650ms (Desktop). Though jQuery is being loadad early in the process, the compiling and executing of the Javascript with additional 300ms takes more time than expected.

I had already tested to reduce jQuery by purging unnecessary modules with Grunt.
But not the file size is the problem (especially with gzip), it is the Javascript compilation and execution.

While (I need to admit) it is really fair at scalability

- Front-Page with 120 Featured Products will load at 2 seconds 
- Front-Page at slow connections and slow mobiles will scale really well and fair and as expected, or even better

But it is not possible to make it **really** quick with only 28 requests (12 Featured Products) for "Desktop" (quick Network, fast CPU)

With jQuery it's _okay quick_, it's scalable when all assets are optimized. But it's not **snappy quick**
It's not as quick as it could be when you want it just modern, slim and quick (feeling).

What I mean with snappy quick is e.g. https://radszene.de (at this example website images are being matched for various resolutions, too).

**I'm fine at the moment as it is. But I think we should keep an eye on that.**

I understand the (good) arguments for the choice of jQuery, nevertheless it would be very interesting to me to be able to exchange for testing with another more slim library. So if anyone likes to cooperate please contact me.

### Links

[Are JavaScript libraries really that bad for performance?](https://gomakethings.com/are-javascript-libraries-really-that-bad-for-performance/)
[The Cost of Javascript](https://medium.com/@addyosmani/the-cost-of-javascript-in-2018-7d8950fbb5d4)

**Reply by drbyte** [reply]
> jQuery is used at zen-cart mainly for image size matching, right?

No, actually it's not used for images, as far as core code goes. It is in the case of some plugins.

**Comment by drbyte** [comment]
## Summary of Zen Cart jQuery Usage

### Admin:
- admin popover effects
- admin interdependent dropdown fields, such as coupon category restrictions 
- admin confirmation modals via jAlert / jTimeout
- admin search
- admin banner graphs via flot library
- date picker 
- bootstrap integration is interdependent with jQuery, at least a base shell of it

### Catalog-side:
- template sidebar column height matching
- template mobile hamburger-menu overlay
- template hide/display of various features on smaller (mobile) screens
- date picker
- address-book country/state/zone selection dependencies
- attribute selection real-time screen updates
- ajax features (built-in, and numerous plugins)

### zc_install:
- all pages use jQuery and ajax

**Comment by drbyte** [comment]
## Why does Zen Cart embed the "full" jQuery library, instead of stripped-down version?

While core Zen Cart uses a relatively small subset of jQuery features, the full library is bundled with the Zen Cart distribution so that plugin don't have distribute ANY jQuery version, but yet still be able to use all of jQuery's capabilities.

Before making that decision, a huge number of contributed plugins bundled severely-outdated jQuery versions, and usually loaded them in a way that replaced the newest version ... and often even loaded it into the browser multiple times ... thus not only increasing load-times but also init times.
